### PR TITLE
Support RSpec 2 and RSpec 3 define matcher messages

### PR DESCRIPTION
**Pull request by simeonwillbanks**
### Problem
#5 was merged to master and dropped RSpec 2 support.  To release and follow [Semantic Versioning](http://semver.org/), `resqutils` would require a major bump.  :cry: 

> MAJOR version when you make incompatible API changes,

### Solution
Use Ruby's [defined? keyword](https://github.com/ruby/ruby/blob/v2_2_3/doc/keywords.rdoc) to figure out which [RSpec define matcher](https://www.relishapp.com/rspec/rspec-expectations/v/3-3/docs/custom-matchers/define-matcher) failure message methods are in scope.

- RSpec 2
  - `failure_message_for_should` 
  - `failure_message_for_should_not`
- RSpec 3 
  - `failure_message`
  - `failure_message_when_negated`

### Testing

https://github.com/simeonwillbanks/rspec-matchers_test was created to test the `defined?` conditional.  It [has a simple module](https://github.com/simeonwillbanks/rspec-matchers_test/blob/master/lib/rspec/matchers_test.rb#L4-L14) where we can [test colors](https://github.com/simeonwillbanks/rspec-matchers_test/blob/master/spec/rspec-matchers_test_spec.rb#L34-L35). 

It defines the [`be_blue` matcher](https://github.com/simeonwillbanks/rspec-matchers_test/blob/master/spec/rspec-matchers_test_spec.rb#L3), and it uses the [same conditionals](https://github.com/simeonwillbanks/rspec-matchers_test/blob/master/spec/rspec-matchers_test_spec.rb#L8-L30) as this Pull Request.

The specs will **always fail**, so we can see the `be_blue` matcher use the `defined?` conditionals.  There are two `puts` statements, so we can see the methods used.

 - [`failure_message_method`](https://github.com/simeonwillbanks/rspec-matchers_test/blob/master/spec/rspec-matchers_test_spec.rb#L14)
 - [`failure_message_when_negated_method`](https://github.com/simeonwillbanks/rspec-matchers_test/blob/master/spec/rspec-matchers_test_spec.rb#L26)

Here's an [RSpec 2 build](https://travis-ci.org/simeonwillbanks/rspec-matchers_test/builds/83396296), and we see it [using the RSpec 2 methods `failure_message_for_should` and `failure_message_for_should_not`](https://travis-ci.org/simeonwillbanks/rspec-matchers_test/builds/83396296#L129-L130).

Here's an [RSpec 3 build](https://travis-ci.org/simeonwillbanks/rspec-matchers_test/builds/83395368), and we it [using the RSpec 3 methods `failure_message` and `failure_message_when_negated`](https://travis-ci.org/simeonwillbanks/rspec-matchers_test/builds/83395368#L130-L131). 

I didn't see how to add tests in `resqutils` without multiple `Gemfile`s.  If there is another way, please let me know.  Thanks!

### CODE REVIEW

**Comment by davetron5000** [comment]
:+1:

I don't know another way to test it, either.  If Travis supports two gemfiles, that might be the way to do it.

**Comment by simeonwillbanks** [comment]
@davetron5000 To test multiple gemfiles, I think we'd have to add a dependency like [WWTD](https://github.com/grosser/wwtd) ([WWTD example usage](http://grosser.it/2014/06/27/testing-multiple-gemfiles-and-rubies-with-wwtd-no-appraisal/)) or [Appraisal](https://github.com/thoughtbot/appraisal).

Should we explore this in a separate Pull Request?

**Comment by davetron5000** [comment]
Looks like travis supports this: http://docs.travis-ci.com/user/languages/ruby/#Testing-against-multiple-versions-of-dependencies 

I think we can just add the second gemfile to `travis.yml`

**Comment by simeonwillbanks** [comment]
@davetron5000 Thanks for the link!  

Unfortunately, the second gemfile wouldn't be installed in development.  It would only be installed on CI.  If we don't want to add a new dependency, this sounds OK.  The suite is pretty small, so CI runs aren't long.  

You can `bundle install --gemfile=gemfiles/rspec2.gemfile` and run the suite, so we could add a note to the `README`.

If we are OK with adding a new dependency, WWTD or Appraisal easily test against all gemfiles in development and on CI.  

If we go this route, I'd like to spike WWTD because [local testing wouldn't change](https://github.com/grosser/wwtd#rake). 

```ruby
# Rakefile
require 'wwtd/tasks'
# run all gemfiles on all rubies
task :default => :wwtd
```

I'm :+1: adding a second gemfile in this Pull Request, and maybe spiking a new dependency in another.  

Thoughts?

**Comment by davetron5000** [comment]
I guess my thinking was that for day-to-day development, we'd just use `Gemfile` and that would override the gemspec to use the latest RSpec.  `Gemfile.rspec2` would exists solely for CI to make sure we haven't broken backwards compatibility (and, as you point out ,you can run it once off if you need it).

*BUT*, wwtd looks pretty cool, and handy, so worth investigating!

File: .travis.yml [hunk 1]
```diff
@@ -2,3 +2,9 @@ language: ruby
 rvm:
   - 2.1.0
   - 2.2.0
+  - 2.0.0
+  - 1.9.3
```
Comment (simeonwillbanks) [inline .travis.yml]: [`Resqutils::StaleWorkersKiller#initialize` uses a Ruby 2 keyword argument](https://github.com/stitchfix/resqutils/blob/master/lib/resqutils/stale_workers_killer.rb#L9), so the Ruby 1.9.3 build fails for [RSpec 3](https://travis-ci.org/stitchfix/resqutils/jobs/90389794) and [RSpec 2](https://travis-ci.org/stitchfix/resqutils/jobs/90389795).

Should we refactor the method or remove 1.9.3 from CI?

File: .travis.yml [hunk 2]
```diff
@@ -2,3 +2,9 @@ language: ruby
...
+  - ruby-head
+gemfile:
+  - Gemfile
+  - gemfiles/rspec2.gemfile
```
Comment (simeonwillbanks) [inline .travis.yml]: Multiple Gemfile Support!


  - On Travis CI, test against RSpec 2 and 3
    - http://docs.travis-ci.com/user/languages/ruby/#Testing-against-multiple-versions-of-dependencies
  - Locally, you can manually test against RSpec 2 and 3
    - RSpec 3
      - `bundle install`
      - `bundle exec rake`
    - RSpec 2
      - `BUNDLE_GEMFILE=gemfiles/rspec2.gemfile bundle install`
      - `BUNDLE_GEMFILE=gemfiles/rspec2.gemfile bundle exec rake`

File: resqutils.gemspec [hunk 3]
```diff
@@ -20,4 +20,5 @@ Gem::Specification.new do |s|
   s.add_runtime_dependency("resque")
   s.add_development_dependency("rake")
   s.add_development_dependency("rspec")
+  s.add_development_dependency("fakeredis")
```
Comment (simeonwillbanks) [inline resqutils.gemspec]: `fakeredis`, so the specs don't require a running Redis instance.

File: spec/resque_matchers_spec.rb [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    end
+
+    it "asserts Resque job has been queued" do
+      Service.new.expensive
+
+      if RSpec::Version::STRING.start_with?("2")
+        "low".should have_job_queued(class: BackgroundJob, args: [])
+      else
+        expect("low").to have_job_queued(class: BackgroundJob, args: [])
+      end
```
Comment (simeonwillbanks) [inline spec/resque_matchers_spec.rb]: Conditionally use RSpec 2 or 3 spec syntax based on RSpec version constant.

File: spec/spec_helper.rb [hunk 5]
```diff
@@ -1,5 +1,7 @@
+require 'fakeredis'
 require 'resque'
 require 'resqutils'
+require 'resqutils/spec'
```
Comment (simeonwillbanks) [inline spec/spec_helper.rb]: Must require custom spec helpers for `spec/resque_matchers_spec.rb` to [`include Resqutils::Spec::ResqueHelpers`](https://github.com/stitchfix/resqutils/pull/6/files#diff-ebad8b958c9407c2b4f044c43486591cR5).

**Comment by simeonwillbanks** [comment]
@davetron5000 Ready for review!

File: resqutils.gemspec (hunk 3 above)
Comment (davetron5000) [inline resqutils.gemspec]: :+1:

File: .travis.yml (hunk 2 above)
Comment (davetron5000) [inline .travis.yml]: Nice!

File: .travis.yml (hunk 1 above)
Comment (davetron5000) [inline .travis.yml]: I'm not sure if we convert this to an option hash, will this be a breaking change to existing callers?

If yes, I'd vote to not make a breaking change and remove 1.9 support.  If no, then refactor to an options hash seems right.

File: spec/resque_matchers_spec.rb (hunk 4 above)
Comment (davetron5000) [inline spec/resque_matchers_spec.rb]: :+1:

**Comment by simeonwillbanks** [comment]
@davetron5000 Any idea why the [build is failing](https://travis-ci.org/stitchfix/resqutils/jobs/90617321#L154)?  It seems random across Ruby versions and RSpec versions. 

It started after refactoring [`StaleWorkersKiller.new` to use an options hash](https://github.com/stitchfix/resqutils/commit/91749a222ba45b24d8fd64868ca415ecd359ef92).

 I added [`RESQUTILS_SECONDS_TO_BE_CONSIDERED_STALE=1`](https://github.com/stitchfix/resqutils/commit/7b70cc7c8e0c6e330c59f23d33a24deafd3f5bdf) for [`StaleWorkers#seconds_to_be_considered_stale_from_env!`](https://github.com/stitchfix/resqutils/blob/master/lib/resqutils/stale_workers.rb#L35).

Locally, I can't reproduce.  I've tried reproducing with the failed build's Ruby and RSpec.  e.g. [Ruby 1.9.3 and RSpec 2](https://travis-ci.org/stitchfix/resqutils/jobs/90613343).

```
~/Code/resqutils support-multiple-rspecs
❯ rbenv local 1.9.3-p551

~/Code/resqutils support-multiple-rspecs*
❯ BUNDLE_GEMFILE=gemfiles/rspec2.gemfile bundle exec rake
/usr/local/var/rbenv/versions/1.9.3-p551/bin/ruby -S rspec ./spec/resque_matchers_spec.rb ./spec/stale_workers_killer_spec.rb ./spec/stale_workers_spec.rb ./spec/worker_killer_job_spec.rb
.................

Finished in 0.03836 seconds
17 examples, 0 failures

Randomized with seed 32279


~/Code/resqutils support-multiple-rspecs*
❯ RESQUTILS_SECONDS_TO_BE_CONSIDERED_STALE=1 BUNDLE_GEMFILE=gemfiles/rspec2.gemfile bundle exec rake
/usr/local/var/rbenv/versions/1.9.3-p551/bin/ruby -S rspec ./spec/resque_matchers_spec.rb ./spec/stale_workers_killer_spec.rb ./spec/stale_workers_spec.rb ./spec/worker_killer_job_spec.rb
.................

Finished in 0.03475 seconds
17 examples, 0 failures

Randomized with seed 57128
```

**Comment by davetron5000** [comment]
I wonder if there's an ordering bug.  It doesn't seem like that spec could ever pass in isolation, but if [this one](https://github.com/stitchfix/resqutils/blob/0cf84d349ddf7c8d6ce638c27b4ff6af1b0a5492/spec/stale_workers_spec.rb#L98) ran first, it would set the value, thus making it work…

**Comment by simeonwillbanks** [comment]
@davetron5000 Good call.  It's ordering.  I can reproduce with the [seed](https://travis-ci.org/stitchfix/resqutils/jobs/90617321#L172) from the failing build.

```
❯ RESQUTILS_SECONDS_TO_BE_CONSIDERED_STALE=1 BUNDLE_GEMFILE=gemfiles/rspec2.gemfile bundle exec rspec spec --seed 22702
...............F.

Failures:

  1) Resqutils::StaleWorkersKiller kill_stale_workers queues a job for all stale workers
     Failure/Error: described_class.new(stale_workers: stale_workers).kill_stale_workers
     RuntimeError:
       You set a stale value of 0 seconds, making all jobs stale; probably not what you want
     # ./lib/resqutils/stale_workers.rb:39:in `seconds_to_be_considered_stale_from_env!'
     # ./lib/resqutils/stale_workers.rb:11:in `initialize'
     # ./lib/resqutils/stale_workers_killer.rb:10:in `new'
     # ./lib/resqutils/stale_workers_killer.rb:10:in `initialize'
     # ./spec/stale_workers_killer_spec.rb:19:in `new'
     # ./spec/stale_workers_killer_spec.rb:19:in `block (3 levels) in <top (required)>'

Finished in 0.03758 seconds
17 examples, 1 failure

Failed examples:

rspec ./spec/stale_workers_killer_spec.rb:21 # Resqutils::StaleWorkersKiller kill_stale_workers queues a job for all stale workers

Randomized with seed 22702
````

I'll refactor the tests to cleanup the `ENV`.  Thanks!

**Comment by simeonwillbanks** [comment]
@davetron5000 :green_heart:

**Pull request merged**